sklearn/feature_selection/tests/test_feature_select.py

```diff
@@ -420,45 +375,63 @@ def test_select_kbest_regression():
     univariate_filter = SelectFpr(f_regression, alpha=0.01)
     X_r = univariate_filter.fit(X, y).transform(X)
-    X_r2 = GenericUnivariateSelect(
-        f_regression, mode='fpr', param=0.01).fit(X, y).transform(X)
-    assert_array_equal(X_r, X_r2)
-    support = univariate_filter.get_support()
     gtruth = np.zeros(20)
     gtruth[:5] = 1
-    assert_array_equal(support[:5], np.ones((5, ), dtype=np.bool))
-    assert_less(np.sum(support[5:] == 1), 3)
+    for mode in ['fdr', 'fpr', 'fwe']:
+        X_r2 = GenericUnivariateSelect(
+            f_regression, mode=mode, param=0.01).fit(X, y).transform(X)
+        assert_array_equal(X_r, X_r2)
+        support = univariate_filter.get_support()
+        assert_array_equal(support[:5], np.ones((5, ), dtype=np.bool))
+        assert_less(np.sum(support[5:] == 1), 3)
 
 
 def test_select_fdr_regression():
     """
-    Test whether the relative univariate feature selection
-    gets the correct items in a simple regression problem
-    with the fdr heuristic
-    """
-    X, y = make_regression(n_samples=200, n_features=20,
-                           n_informative=5, shuffle=False, random_state=0)
-
-    univariate_filter = SelectFdr(f_regression, alpha=0.01)
-    X_r = univariate_filter.fit(X, y).transform(X)
-    X_r2 = GenericUnivariateSelect(
-        f_regression, mode='fdr', param=0.01).fit(X, y).transform(X)
-    assert_array_equal(X_r, X_r2)
-    support = univariate_filter.get_support()
-    gtruth = np.zeros(20)
-    gtruth[:5] = 1
-    assert_array_equal(support, gtruth)
+    Test that fdr heuristic actually has low FDR.
+    """
+    def single_fdr(alpha, n_informative, random_state):
+        n_features = 5 * n_informative
+
+        X, y = make_regression(n_samples=150, n_features=n_features,
+                               n_informative=n_informative, shuffle=False,
+                               random_state=random_state, noise=10)
+
+        univariate_filter = SelectFdr(f_regression, alpha=alpha)
+        X_r = univariate_filter.fit(X, y).transform(X)
+        X_r2 = GenericUnivariateSelect(
+            f_regression, mode='fdr', param=alpha).fit(X, y).transform(X)
+        assert_array_equal(X_r, X_r2)
+        support = univariate_filter.get_support()
+        num_false_positives = np.sum(support[n_informative:] == 1)
+        num_true_positives = np.sum(support[:n_informative] == 1)
+
+        false_discovery_rate = float(num_false_positives) / \
+            (num_true_positives + num_false_positives)
+        return false_discovery_rate
+
+    for alpha in [0.01, 0.05, 0.1, 0.2]:
+        for n_informative in [5, 10, 20]:
+            # We have that
+            #  FDR = E(FP / (TP + FP)) <= alpha
+            false_discovery_rate = np.mean([single_fdr(alpha, n_informative,
+                                                       random_state) for
+                                            random_state in range(40)])
+            assert_true(false_discovery_rate <= alpha)
```

Comment:
Actually the goal of the [Benjamini-Hochberg procedure](http://en.wikipedia.org/wiki/False_discovery_rate#Benjamini.E2.80.93Hochberg_procedure) is to make alpha an upper bound of the expected FDR. So this test is correct.

I will submit a PR to clarify the docstring and catch warnings raised by this test.